File: ss_ts/src/scheme.ts
import * as fs from "fs";
import * as readline from "readline";

// Evaluator
type Lambda = (args: Val[], env: Env) => Val;
type Val = number | boolean | Lambda;

function vprint(val: Val) {
  if (typeof val === "number") {
    console.log(val);
  } else if (typeof val === "function") {
    console.log("<lambda function>");
  } else if (val === true) {
    console.log("#t");
  } else {
    console.log("#f");
  }
}

class Env {
  parent: Env | null;
  bindings: Map<string, Val>;

  constructor(parent: Env | null, bindings: Map<string, Val>) {
    this.parent = parent;
    this.bindings = bindings;
  }

  find(name: string): Env | null {
    if (this.bindings.has(name)) {
      return this;
    } else if (this.parent !== null) {
      return this.parent.find(name);
    } else {
      return null;
    }
  }

  get(name: string): Val {
    const env = this.find(name);
    if (env !== null) {
      return env.bindings.get(name) as Val;
    } else {
      throw Error(`${name} is not defined`);
    }
  }

  set(name: string, val: Val): void {
    this.bindings.set(name, val);
  }
}

function _define(expr: Expr, env: Env) {
  if (!Array.isArray(expr) || expr.length !== 3) {
    throw Error("Syntax error: 'define' expects exactly 2 arguments (name and value).");
  }
  const name = expr[1];
  if (typeof name !== "string") {
    throw Error(`Syntax error: 'define' expects a symbol as the first argument, but got ${JSON.stringify(name)}`);
  }
  const value = evaluate(expr[2] as Expr, env);
  env.set(name, value);
  return;
}

function _if(expr: Expr, env: Env): Val {
  if (!Array.isArray(expr) || (expr.length !== 3 && expr.length !== 4)) {
    throw Error("Syntax error: 'if' expects 2 or 3 arguments.");
  }
  const test = evaluate(expr[1] as Expr, env);
  if (typeof test !== "boolean") {
    throw Error("Type error: 'if' test expression must evaluate to a boolean.");
  }
  if (test) {
    return evaluate(expr[2] as Expr, env);
  } else {
    if (expr.length === 4) {
      return evaluate(expr[3] as Expr, env);
    } else {
      return false;
    }
  }
}

function _and(expr: Expr, env: Env): Val {
  if (!Array.isArray(expr) || expr.length < 2) {
    throw Error("Syntax error: 'and' expects at least one argument.");
  }
  let result: Val = true;
  for (let i = 1; i < expr.length; i++) {
    result = evaluate(expr[i] as Expr, env);
    if (result === false) {
      return false;
    }
  }
  return result;
}

function _or(expr: Expr, env: Env): Val {
  if (!Array.isArray(expr) || expr.length < 2) {
    throw Error("Syntax error: 'or' expects at least one argument.");
  }
  let result: Val = false;
  for (let i = 1; i < expr.length; i++) {
    result = evaluate(expr[i] as Expr, env);
    if (result !== false) {
      return result;
    }
  }
  return false;
}

function _lambda(expr: Expr, env: Env): Val {
  if (!Array.isArray(expr) || expr.length < 3) {
    throw Error("Syntax error: 'lambda' expects at least 2 arguments (parameters and body).");
  }
  const params = expr[1];
  if (!Array.isArray(params) || !params.every((p) => typeof p === "string")) {
    throw Error("Syntax error: 'lambda' parameters must be a list of symbols.");
  }
  const body = expr.slice(2);
  return function (args: Val[], subenv: Env): Val {
    if (args.length !== params.length) {
      throw Error(`Argument count mismatch: Lambda function expected ${params.length} arguments but got ${args.length}.`);
    }
    const lambdaEnv = new Env(env, new Map<string, Val>());
    for (let i = 0; i < params.length; i++) {
      lambdaEnv.set(params[i] as string, args[i] as Val);
    }
    let result: Val = false;
    for (const expr of body) {
      result = evaluate(expr as Expr, lambdaEnv);
    }
    return result;
  };
}

const special_forms = new Map<string, (args: Expr[], env: Env) => Val>([
  ["if", _if],
  ["lambda", _lambda],
  ["and", _and],
  ["or", _or],
]);

function evaluate(expr: Expr, env: Env): Val {
  if (typeof expr === "number" || typeof expr === "boolean") {
    return expr;
  }
  if (typeof expr === "string") {
    return env.get(expr);
  }
  if (!Array.isArray(expr) || expr.length === 0) {
    throw Error("Evaluation error: Cannot evaluate an empty list.");
  }

  const form = expr[0] as Expr;
  if (typeof form === "string" && special_forms.has(form)) {
    return special_forms.get(form)!(expr, env);
  } else {
    const func = evaluate(form, env);
    if (typeof func !== "function") {
      throw Error(`Type error: Expression starts with non-callable value: ${JSON.stringify(form)}`);
    }

    const args = expr.slice(1).map((arg) => evaluate(arg as Expr, env));
    return func(args, env);
  }
}

// Parser
type Sym = string;
type Expr = number | boolean | Sym | Expr[];

function readNumber(code: string): [number, string] {
  const numberPattern = /^[-+]?\d+(\.\d+)?/;
  const match = code.match(numberPattern);

  if (match) {
    const numberStr = match[0];
    const parsedNumber = parseFloat(numberStr);
    const remain = code.substring(numberStr.length);
    return [parsedNumber, remain];
  }

  throw Error("Parser error: Invalid number format.");
}

function readSym(code: string): [string, string] {
  for (let i = 1; i <= code.length; i++) {
    if (code[i]?.trim().length === 0 || code[i] === ")" ) {
      const sym = code.slice(0, i);
      return [sym, code.slice(i)];
    }
  }
  return [code, ""]
}

function readExpr(code: string): [Expr, string] {
  const code_trimmed = code.trim();
  if (code_trimmed.length === 0) {
    throw Error("Parser error: Unexpected end of input.");
  }
  const first_char = code_trimmed[0] as string;
  if (first_char === "(") {
    let remain = code_trimmed.slice(1);
    let exprs = [] as Expr[];
    while (remain[0] != ")") {
      let expr;
      [expr, remain] = readExpr(remain);
      exprs.push(expr);
    }
    return [exprs, remain.slice(1)];
  } else if (first_char === "#") {
    if (code_trimmed.length < 2 || (code_trimmed[1] !== "t" && code_trimmed[1] !== "f")) {
      throw Error("Parser error: Invalid boolean literal. Use #t or #f.");
    }
    return [code_trimmed[1] == "t" ? true : false, code_trimmed.slice(2)];
  } else if (/^[+-]?\d$/.test(first_char) || (
    (first_char === "-" || first_char === "+") && code_trimmed.length > 1 &&
    /^\d/.test(code_trimmed[1] as string)
  )) {
    return readNumber(code_trimmed);
  } else {
    return readSym(code_trimmed);
  }
}


const globalEnv = new Env(null, new Map<string, Lambda>([
  ["+", (args: Val[], _: Env) => args.reduce((a, b) => (a as number) + (b as number))],
  ["*", (args: Val[], _: Env) => args.reduce((a, b) => (a as number) * (b as number))],
  ["-", (args: Val[], _: Env) => args.reduce((a, b) => (a as number) - (b as number))],
  ["/", (args: Val[], _: Env) => args.reduce((a, b) => (a as number) / (b as number))],
  ["=", (args: Val[], _: Env) => args.reduce((a, b) => a === b)],
  [">", (args: Val[], _: Env) => args.reduce((a, b) => a > b)],
  ["<", (args: Val[], _: Env) => args.reduce((a, b) => a < b)],
  [">=", (args: Val[], _: Env) => args.reduce((a, b) => a >= b)],
  ["<=", (args: Val[], _: Env) => args.reduce((a, b) => a <= b)],
  ["not", (args: Val[], _: Env) => {
    if (args.length !== 1 || typeof args[0] !== "boolean") {
      throw Error(`Type error: 'not' expects a single boolean argument, but got ${args.length} arguments.`);
    }
    return !args[0];
  }],
]));

function file_mode(test_file: string) {
  const code = fs.readFileSync(test_file, "utf-8");
  let remain = code.trim();
  remain = remain.replace(/;.*$/gm, "").trim();

  // console.log(`Running file: ${test_file}`);
  // console.log(`Remain: ${remain}`);

  while (remain.trim().length != 0) {
    let expr;
    [expr, remain] = readExpr(remain);
    // console.log(expr);
    if (Array.isArray(expr) && expr[0] === "define") {
      _define(expr, globalEnv);
    } else {
      vprint(evaluate(expr, globalEnv));
    }
  }
}

function repl_mode() {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    prompt: "scheme> ",
  });
  rl.on("line", (line) => {
    if (line.trim().length === 0) {
      rl.prompt();
      return;
    }
    const [expr, remain] = readExpr(line);
    if (remain.trim().length != 0) {
      console.log("Only one expression per line!");
    }
    if (Array.isArray(expr) && expr[0] === "define") {
      _define(expr, globalEnv);
    } else {
      vprint(evaluate(expr, globalEnv));
    }
    rl.prompt();
  }).on("close", () => {
    console.log("Exiting...");
    process.exit(0);
  });
  rl.prompt();
}

// process.argv[1] = "../tests/basic_arithmetic.ss"; // For testing in VSCode
if (process.argv.length >= 3) {
  file_mode(process.argv[2] as string);
} else {
  repl_mode();
}

File: ss_go/scheme.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
	"unicode"
)

// ==========================================
// 1. 类型定义与环境 (Types & Environment)
// ==========================================

type Env struct {
	parent *Env
	store  map[string]Val
}

func NewEnv(parent *Env) *Env {
	return &Env{parent: parent, store: make(map[string]Val)}
}

func (e *Env) Get(name string) (Val, bool) {
	if v, ok := e.store[name]; ok {
		return v, true
	}
	if e.parent != nil {
		return e.parent.Get(name)
	}
	return nil, false
}

func (e *Env) Set(name string, val Val) {
	e.store[name] = val
}

type Val interface {
	String() string // Go 惯用 String() 而非 repr()
}

type (
	Number  float64
	Bool    bool
	Symbol  string
	List    []Expr
	Builtin func([]Val) (Val, error) // 简化原生函数签名
	Lambda  struct {
		params []string
		body   Expr
		env    *Env
	}
)

func (n Number) String() string { return strconv.FormatFloat(float64(n), 'f', -1, 64) }
func (b Bool) String() string {
	if b {
		return "#t"
	}
	return "#f"
}
func (s Symbol) String() string { return string(s) }
func (l List) String() string {
	parts := make([]string, len(l))
	for i, e := range l {
		parts[i] = fmt.Sprintf("%v", e) // 简化打印
	}
	return "(" + strings.Join(parts, " ") + ")"
}
func (b Builtin) String() string { return "<builtin>" }
func (l Lambda) String() string  { return "<lambda>" }

// ==========================================
// 2. 表达式与求值 (Expression & Evaluation)
// ==========================================

type Expr interface {
	Eval(env *Env) (Val, error)
}

// 原子类型自求值
func (n Number) Eval(env *Env) (Val, error) { return n, nil }
func (b Bool) Eval(env *Env) (Val, error)   { return b, nil }
func (s Symbol) Eval(env *Env) (Val, error) {
	if v, ok := env.Get(string(s)); ok {
		return v, nil
	}
	return nil, fmt.Errorf("undefined symbol: %s", s)
}

// 列表求值（核心逻辑）
func (l List) Eval(env *Env) (Val, error) {
	if len(l) == 0 {
		return nil, fmt.Errorf("cannot evaluate empty list")
	}

	// 1. 检查是否为 Special Form (if, define, lambda, etc.)
	// 这里为了简化，假设 Special Form 的第一个元素必须是 Symbol
	if headSym, ok := l[0].(Symbol); ok {
		if handler, ok := specialForms[string(headSym)]; ok {
			return handler(l[1:], env)
		}
	}

	// 2. 函数调用：先求值第一个元素
	fnVal, err := l[0].Eval(env)
	if err != nil {
		return nil, err
	}

	// 3. 求值所有参数
	args := make([]Val, 0, len(l)-1)
	for _, expr := range l[1:] {
		arg, err := expr.Eval(env)
		if err != nil {
			return nil, err
		}
		args = append(args, arg)
	}

	// 4. 执行函数
	switch fn := fnVal.(type) {
	case Builtin:
		return fn(args)
	case Lambda:
		if len(args) != len(fn.params) {
			return nil, fmt.Errorf("arity mismatch: expected %d args, got %d", len(fn.params), len(args))
		}
		// 创建闭包环境
		newEnv := NewEnv(fn.env)
		for i, param := range fn.params {
			newEnv.Set(param, args[i])
		}
		return fn.body.Eval(newEnv)
	default:
		return nil, fmt.Errorf("not a function: %s", l[0])
	}
}

// ==========================================
// 3. Special Forms & Builtins (统一管理)
// ==========================================

// SpecialForm 处理器接收未求值的参数 AST
type SpecialForm func(args []Expr, env *Env) (Val, error)

var specialForms = map[string]SpecialForm{
	"define": func(args []Expr, env *Env) (Val, error) {
		if len(args) != 2 {
			return nil, fmt.Errorf("define takes 2 args")
		}
		sym, ok := args[0].(Symbol)
		if !ok {
			return nil, fmt.Errorf("define first arg must be symbol")
		}
		val, err := args[1].Eval(env)
		if err != nil {
			return nil, err
		}
		env.Set(string(sym), val)
		return nil, nil // define 返回 nil
	},
	"if": func(args []Expr, env *Env) (Val, error) {
		if len(args) != 3 {
			return nil, fmt.Errorf("if takes 3 args")
		}
		cond, err := args[0].Eval(env)
		if err != nil {
			return nil, err
		}
		// 只有 #f 是 false，其他都是 true
		if b, ok := cond.(Bool); ok && !bool(b) {
			return args[2].Eval(env)
		}
		return args[1].Eval(env)
	},
	"lambda": func(args []Expr, env *Env) (Val, error) {
		if len(args) != 2 {
			return nil, fmt.Errorf("lambda takes 2 args")
		}
		paramsNode, ok := args[0].(List)
		if !ok {
			return nil, fmt.Errorf("lambda params must be a list")
		}
		params := make([]string, len(paramsNode))
		for i, p := range paramsNode {
			sym, ok := p.(Symbol)
			if !ok {
				return nil, fmt.Errorf("param must be symbol")
			}
			params[i] = string(sym)
		}
		return Lambda{params: params, body: args[1], env: env}, nil
	},
	"and": func(args []Expr, env *Env) (Val, error) {
		if len(args) != 2 {
			return nil, fmt.Errorf("and takes 2 args")
		}
		cond1, err := args[0].Eval(env)
		if err != nil {
			return nil, err
		}
		if cond1b, ok := cond1.(Bool); ok {
			if !cond1b {
				return cond1, nil
			}
			cond2, err := args[1].Eval(env)
			if err != nil {
				return nil, err
			}
			if cond2b, ok := cond2.(Bool); ok {
				return cond2b, nil
			}
			return nil, fmt.Errorf("and arg must be a bool")
		}
		return nil, fmt.Errorf("and arg must be a bool")
	},
	"or": func(args []Expr, env *Env) (Val, error) {
		if len(args) != 2 {
			return nil, fmt.Errorf("or takes 2 args")
		}
		cond1, err := args[0].Eval(env)
		if err != nil {
			return nil, err
		}
		if cond1b, ok := cond1.(Bool); ok {
			if cond1b {
				return cond1, nil
			}
			cond2, err := args[1].Eval(env)
			if err != nil {
				return nil, err
			}
			if cond2b, ok := cond2.(Bool); ok {
				return cond2b, nil
			}
			return nil, fmt.Errorf("or arg must be a bool")
		}
		return nil, fmt.Errorf("or arg must be a bool")
	},
}

// 辅助函数：将 Number 转换逻辑抽离
func assertNums(args []Val) ([]float64, error) {
	nums := make([]float64, len(args))
	for i, arg := range args {
		n, ok := arg.(Number)
		if !ok {
			return nil, fmt.Errorf("expected number, got %T", arg)
		}
		nums[i] = float64(n)
	}
	return nums, nil
}

// 辅助函数：创建二元数值操作符 (减少重复代码)
func binaryNumOp(op func(a, b float64) float64) Builtin {
	return func(args []Val) (Val, error) {
		nums, err := assertNums(args)
		if err != nil {
			return nil, err
		}
		if len(nums) < 2 {
			// 这里简单处理，严格来说 - 可以是一元操作符
			return nil, fmt.Errorf("arithmetic op requires at least 2 args")
		}
		res := nums[0]
		for _, n := range nums[1:] {
			res = op(res, n)
		}
		return Number(res), nil
	}
}

// 辅助函数：创建比较操作符
func compareOp(op func(a, b float64) bool) Builtin {
	return func(args []Val) (Val, error) {
		if len(args) != 2 {
			return nil, fmt.Errorf("comparison requires 2 args")
		}
		nums, err := assertNums(args)
		if err != nil {
			return nil, err
		}
		return Bool(op(nums[0], nums[1])), nil
	}
}

func loadBuiltins(env *Env) {
	env.Set("+", binaryNumOp(func(a, b float64) float64 { return a + b }))
	env.Set("-", binaryNumOp(func(a, b float64) float64 { return a - b }))
	env.Set("*", binaryNumOp(func(a, b float64) float64 { return a * b }))
	env.Set("/", binaryNumOp(func(a, b float64) float64 { return a / b }))
	env.Set("=", compareOp(func(a, b float64) bool { return a == b }))
	env.Set("<", compareOp(func(a, b float64) bool { return a < b }))
	env.Set("<=", compareOp(func(a, b float64) bool { return a <= b }))
	env.Set(">=", compareOp(func(a, b float64) bool { return a >= b }))
	env.Set(">", compareOp(func(a, b float64) bool { return a > b }))
	env.Set("not", Builtin(func(args []Val) (Val, error) {
		if len(args) != 1 {
			return nil, fmt.Errorf("not takes 1 args")
		}
		if val, ok := args[0].(Bool); ok {
			return !val, nil
		}
		return nil, fmt.Errorf("not takes bool arg")
	}))
}

// ==========================================
// 4. 解析器 (Parser) - 使用 Tokenizer 改进
// ==========================================

func readExpr(code string) (Expr, string, error) {
	code = strings.TrimSpace(code)

	if len(code) == 0 {
		return nil, "", errors.New("unexpected EOF")
	}

	first := code[0]
	if first == '(' {
		remain := code[1:]
		list := List{}
		for {
			remain = strings.TrimSpace(remain)
			if len(remain) == 0 {
				return nil, "", errors.New("unbalanced parenthesis: missing ')'")
			}
			if remain[0] == ')' {
				return list, remain[1:], nil
			}
			expr, remainNext, err := readExpr(remain)
			if err != nil {
				return nil, "", err
			}
			list = append(list, expr)
			remain = remainNext
		}
	}

	nextParamsIdx := strings.IndexFunc(code, func(r rune) bool {
		return unicode.IsSpace(r) || r == ')' || r == '('
	})

	var token string
	var nextRemain string

	if nextParamsIdx == -1 {
		token = code
		nextRemain = ""
	} else {
		token = code[:nextParamsIdx]
		nextRemain = code[nextParamsIdx:]
	}
	if token == "#t" {
		return Bool(true), nextRemain, nil
	} else if token == "#f" {
		return Bool(false), nextRemain, nil
	} else if num, err := strconv.ParseFloat(token, 64); err == nil {
		return Number(num), nextRemain, nil
	} else {
		return Symbol(token), nextRemain, nil
	}
}

func fileMode(testFile string, env *Env) {

	// 1. 读取整个文件
	bytes, err := os.ReadFile(testFile)
	if err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		return
	}

	code := string(bytes)
	remain := strings.TrimSpace(code)

	// 2. 正则去除注释 (; 到行尾)，对应 TS 的 .replace(/;.*$/gm, "")
	// (?m) 开启多行模式
	re := regexp.MustCompile(`(?m);.*$`)
	remain = re.ReplaceAllString(remain, "")
	remain = strings.TrimSpace(remain)

	// fmt.Printf("Running file: %s\n", testFile)
	// 3. 循环消费字符串

	for len(strings.TrimSpace(remain)) > 0 {
		var expr Expr
		var err error
		expr, remain, err = readExpr(remain)
		if err != nil {
			fmt.Printf("Syntax Error: %v\n", err)
			break
		}

		res, err := expr.Eval(env)

		if err != nil {
			fmt.Printf("Error: %v\n", err)
		} else if res != nil {
			fmt.Println(res.String())
		}
	}
}

// 核心逻辑：REPL 模式
func replMode(env *Env) {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Print("scheme> ")

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			fmt.Print("scheme> ")
			continue
		}

		// 解析当前行
		expr, remain, err := readExpr(line)
		if err != nil {
			fmt.Printf("Syntax Error: %v\n", err)
			fmt.Print("scheme> ")
			continue
		}

		if len(strings.TrimSpace(remain)) != 0 {
			fmt.Println("Only one expression per line!")
		}

		res, err := expr.Eval(env)

		if err != nil {
			fmt.Printf("Error: %v\n", err)
		} else if res != nil {
			fmt.Println(res.String())
		}

		fmt.Print("scheme> ")
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintln(os.Stderr, "reading standard input:", err)
	}
	fmt.Println("Exiting...")
}

// ==========================================
// 5. Main Loop
// ==========================================

// todo：scheme 添加 future / promise 并发支持
func main() {
	env := NewEnv(nil)
	loadBuiltins(env)

	if len(os.Args) > 1 {
		// 直接调用封装好的 runFile
		fileMode(os.Args[1], env)
	} else {
		// REPL 模式 (逻辑也可以复用 tokenize)
		replMode(env)
	}
}
